x-pack/filebeat/input/awss3/acker.go

```diff
@@ -0,0 +1,78 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package awss3
+
+import (
+	"context"
+	"sync"
+
+	"github.com/elastic/beats/v7/libbeat/beat"
+	"github.com/elastic/beats/v7/libbeat/common/acker"
+)
+
+// eventACKTracker tracks the publishing state of S3 objects. Specifically
+// it tracks the number of message acknowledgements that are pending from the
+// output. It can be used to wait until all ACKs have been received for one or
+// more S3 objects.
+type eventACKTracker struct {
+	sync.Mutex
+	pendingACKs int64
+	ctx         context.Context
+	cancel      context.CancelFunc
+}
+
+func newEventACKTracker(ctx context.Context) *eventACKTracker {
+	ctx, cancel := context.WithCancel(ctx)
+	return &eventACKTracker{ctx: ctx, cancel: cancel}
+}
+
+// Add increments the number of pending ACKs by the specified amount.
+func (a *eventACKTracker) Add(messageCount int64) {
+	a.Lock()
+	a.pendingACKs++
+	a.Unlock()
+}
+
+// ACK decrements the number of pending ACKs.
+func (a *eventACKTracker) ACK() {
+	a.Lock()
+	defer a.Unlock()
+
+	if a.pendingACKs <= 0 {
+		panic("misuse detected: negative ACK counter")
+	}
+
+	a.pendingACKs--
+	if a.pendingACKs == 0 {
+		a.cancel()
+	}
+}
+
+// Wait waits for the number of pending ACKs to be zero.
```

Comment:
I'm probably missing something, but does this actually wait until the pending acks are 0?

Reply:
I added a comment to explain the function. The small bit logic there is to handle a special case of no events.